internal/mode/static/handler.go

```diff
@@ -158,44 +171,53 @@ func (h *eventHandlerImpl) HandleEventBatch(ctx context.Context, logger logr.Log
 	h.cfg.statusUpdater.Update(ctx, buildGatewayAPIStatuses(graph, gwAddresses, nginxReloadRes))
 }
 
-func (h *eventHandlerImpl) handleEvent(ctx context.Context, logger logr.Logger, event interface{}) {
-	switch e := event.(type) {
-	case *events.UpsertEvent:
-		switch obj := e.Resource.(type) {
-		case *ngfAPI.NginxGateway:
-			h.updateControlPlaneAndSetStatus(ctx, logger, obj)
-		case *apiv1.Service:
-			podConfig := h.cfg.gatewayPodConfig
-			if obj.Name == podConfig.ServiceName && obj.Namespace == podConfig.Namespace {
-				gwAddresses, err := getGatewayAddresses(ctx, h.cfg.k8sClient, obj, h.cfg.gatewayPodConfig)
-				if err != nil {
-					logger.Error(err, "Setting GatewayStatusAddress to Pod IP Address")
-				}
-				h.cfg.statusUpdater.UpdateAddresses(ctx, gwAddresses)
+func (h *eventHandlerImpl) handleUpsertEvent(ctx context.Context, logger logr.Logger, event *events.UpsertEvent) {
```

Comment:
we now have 3 cases when we watch for changes to a single resource:
- NginxGateway resource
- Fronting Service
- NIM Secret

I think we should come up with some mechanism that allow our components to subscribe to changes to certain resources by their Kind and namespace/name, rather than having this logic in the handler.

This logic should probably part of our EventLoop. For example:
```
func (el *EventLoop) FilterChangesFor(
	objectType client.Object,
	nsname types.NamespacedName,
) <-chan interface{} {
	//	
}
```

As a result, whenever the even loop processes a change for a filtered resource, it will publish it to that returned channel -- https://github.com/nginxinc/nginx-gateway-fabric/blob/074b5269687a2a3a2606ef44bbcb3c86fada6c56/internal/framework/events/event.go#L12-L18

Reply:
I like the idea, let me think about it a bit more.

Reply:
Though one issue right now is that the event loop is agnostic and used for both the provisioner and static mode.

Reply:
Maybe we initialize the event loop with a filter that contains the object and a callback function (defined on the handler). Whenever the event loop encounters a special object, it calls the callback. In the default case, it calls HandleEventBatch().

Reply:
> Whenever the event loop encounters a special object, it calls the callback. In the default case, it calls HandleEventBatch().

it probably should call both, so that the filter object could also be processed in the batch.

> Maybe we initialize the event loop with a filter that contains the object and a callback function (defined on the handler).

probably many filters, right? Once per object?

Reply:
> probably many filters, right? Once per object?

Yes, many filters

> it probably should call both, so that the filter object could also be processed in the batch.

Why? If the object doesn't directly generate any nginx config (like NginxGateway CRD or usage secret), what benefit is there to calling HandleEventBatch?

Reply:
> Why? If the object doesn't directly generate any nginx config (like NginxGateway CRD or usage secret), what benefit is there to calling HandleEventBatch?

Edge cases when the same resource also referenced by Gateway API objects. HTTPRoute referencing a Service, Gateway a Secret. 
So that we handle those edge cases appropriately. 

> Though as I'm implementing this more, it may make more sense for HandleEventBatch to call the filter function...since it parses and determines the event type.

It feels to me more like a framework-related functionality, rather than the feature of the static mode. 
Another point - I don't think we need batching for those resource changes - it is unlikely they will change multiple times per seconds.

Reply:
As I worked through this, what I found is that if the event loop was to determine the object and then call its filter, then the event loop would now have to parse the event. HandleEventBatch already parses the event and makes a decision based on the object, so we would be duplicating efforts here. It also makes sense to me that the handler still makes the decision on what to do with the object, since that's its purpose.

Regarding batching, we can still process Gateway resources and special resources as one batch to reduce the overall handler calls. The handler parses through each event/object, and determines what to do with it.